### ISSUE TITLE
Support --credential-spec in docker compose / docker stack

### ISSUE THREAD

**Issue by BrandonRoyal** [moby/moby#32053]
**Description**
To support use of Windows Domain Identities and [Windows Credential Spec](https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/manage-serviceaccounts), we're looking for support for `--credential-spec` in `docker compose`/`docker stack deploy`

 **Additional environment details (AWS, VirtualBox, physical, etc.):**
Windows Server 2016

cc @sixeyed @friism

**Comment by pcgeek86** [comment]
We definitely will need support for this to help deploy Docker Swarm services with Group Managed Services Accounts (gMSA) for Windows Server Containers. We need to deploy containers with Active Directory domain access, and `--credential-spec` is how it's done.

Are there any workarounds for this, to unblock?

Cheers,
Trevor Sullivan
Amazon Web Services (AWS)

**Comment by pcgeek86** [comment]
The proper parameter to launch a Windows Server Container under a gMSA would be `--security-opt "credentialspec=..."`

https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-containers/manage-serviceaccounts

Cheers,
Trevor Sullivan

**Comment by jhowardmsft** [comment]
I think there's a number of advantages to continue using `security-opt`. Particularly, avoiding deprecation, that it's been documented for many months on MSDN/TechNet, and not having to include a new `hostconfig` field. Then, of course, in the migration, coping with both situations. JMHI. I don't have a strong opinion though.  The `--credentialspec` was only removed as it was an oversight. It's never been wired up - purely an oversight in master that was never picked up in code review or my own validation.

@johnstep who is looking into this.

**Comment by johnstep** [comment]
We are planning to add `--credentialspec` to `docker service create` and want to add it to `docker run`, for consistency, while still supporting `--security-opt` for `docker run`. Longer term, we would like to deprecate the `--security-opt` approach, though.

**Comment by aluzzardi** [comment]
@jhowardmsft If I understand correctly, credential spec is not a security feature right?

**Comment by jhowardmsft** [comment]
In the sense that it allows containers to access domain joined resources, then yes. @patricklang can probably point you to some documentation on what it does under the covers once the container has the credential spec.

**Comment by BrandonRoyal** [comment]
@aluzzardi @johnstep now that we have `--credential-spec` in `docker service create` could we discuss prioritizing this?

**Comment by cpuguy83** [comment]
@BrandonRoyal #32393

**Comment by Vacant0mens** [comment]
any projection on when this will get released to docker EE?

**Comment by cpuguy83** [comment]
@Vacant0mens 17.06.

Closing since this is added.

**Comment by friism** [comment]
@Vacant0mens as @cpuguy83 mentioned this is going to be in the 17.06 release. If you want, you can test release clients from here: https://download.docker.com/win/static/test/x86_64/

EE will follow later

**Comment by Vacant0mens** [comment]
any word on when 17.06 will be official to EE?

**Comment by friism** [comment]
@Vacant0mens no, nothing firm. Usually EE ships 3-5 weeks after CE.

**Comment by Vacant0mens** [comment]
I apologize @friism, I'm not very familiar with the release process yet. I understand 17.06 for CE is in Edge right now, correct? About how long will it stay there before going to Stable? 
And then once it's in Stable, is that when the 3-5 weeks start for getting it to EE? Or is 17.06 in Edge for both CE and EE currently but will stay in Edge for EE 3-5 weeks longer?

**Comment by friism** [comment]
@Vacant0mens 17.06 CE GA is imminent (the current rc5 is hopefully the final one before GA).

Then we'll start working on 17.06 EE.

**Comment by Vacant0mens** [comment]
@friism  I sent you an email. Hope that's okay.

**Comment by friism** [comment]
@Vacant0mens 

IIRC this is how the compose file should look (this took me some time to figure out too :-)):

```
version: "3.3"
services:
  test:
    image: microsoft/windowsservercore
    credential_spec:
      file: "WebApplication1.json"
    command: powershell -c Sleep 3600
    deploy:
      endpoint_mode: dns-rr
```

**Comment by Vacant0mens** [comment]
Thanks!

**Comment by artisticcheese** [comment]
This is only supported in "swarm" mode and not in "compose" per instructions and error message. Why is it not supported in "compose"?

```

**Comment by thaJeztah** [comment]
@artisticcheese docker compose does not create swarm services, so there’s no need to pass this option; compose will pull using your local credentials, but doesn’t store those credentials in the service definition

**Comment by artisticcheese** [comment]
What do you mean it does not create services? If I run `docker-compose up` it will create those 2 services on my local machine, is not it?

**Comment by thaJeztah** [comment]
It’s starting a container for those services, but not a swarm-mode service. i.e., it does not deploy to a swarm cluster

**Comment by artisticcheese** [comment]
Correct, so why I can not use credentialspec in such cases?

**Comment by thaJeztah** [comment]
Oh, gosh, ignore me, I was confusing `—credentials-spec` for `—with-registry-auth`; yes that should be possible; but something to be implemented in docker compose; https://github.com/docker/compose

**Comment by jorisscheppers** [comment]
@artisticcheese @thaJeztah According to this, the support for the --credentials-spec option was added in compose v3.3: https://docs.docker.com/compose/compose-file/#credential_spec

**Comment by artisticcheese** [comment]
Yes but it's only for swarm mode per error message.

**Comment by jorisscheppers** [comment]
So then either there's a bug in compose or the documentation is wrong/out-of-date. If --credentials-spec is only supported in swarm-mode, it should be noted in the documentation. Under 'deploy' (https://docs.docker.com/compose/compose-file/#deploy) it is explicitly mentioned that the option only works in swarm-mode:

> This only takes effect when deploying to a swarm with docker stack deploy, and is ignored by docker-compose up and docker-compose run

**Comment by kosxe** [comment]
Has this ever been fixed? I have just tried using "credential_spec" with docker-compose up and got the same error message: WARNING: Some services (docker_test) use the 'credential_spec' key, which will be ignored. Compose does not support 'credential_spec' configuration - use `docker stack deploy` to deploy to a swarm.

**Comment by thaJeztah** [comment]
See https://github.com/docker/compose/issues/5488

**Comment by kosxe** [comment]
Although, that does confirm, that it does not work, no where has anyone responded on why it is not supported in non-swarm.

**Comment by markdwags** [comment]
@kosxe -- what did you use to get this working?

I'm still getting an error that says "credential_spec" is unsupported.

**Comment by kosxe** [comment]
I did something like this:
services:
  myapp:
    image: my.app
    build:
      context: .
      dockerfile: my.app/Dockerfile
    tty: true
    security_opt:
      - credentialspec=file://myGMSA.json

**Comment by artisticcheese** [comment]
Did this work for you without swarm?

**Comment by kosxe** [comment]
@artisticcheese yes, it does.

**Comment by pkharris** [comment]
This needs to be supported using docker stack deploy.  Also, why do we use "docker stack deploy" to deploy to swarm?  Can there not be a section added to yaml for use with docker-compose?  I definitely DO NOT like there are two yaml specs for orchestration.

**Comment by Vacant0mens** [comment]
@pkharris this issue was closed because both `docker-compose` and `docker stack` support it, as of 4 years ago.

If you've done the recommended steps to use the credential spec option properly and it's not working for some reason, feel free to open a new issue and include more details/errors.

There are two yaml specs for orchestration because they serve different purposes. Feel free to open an issue about it if you'd like to make a push for combining them somehow.
